src/chunkserver/chunkserver-common/chunk_trash_manager_impl.cc

```
@@ -0,0 +1,297 @@
+		if (fs::is_regular_file(file) && isTrashPath(file.path().string())) {
+			const std::string filename = file.path().filename().string();
+			const std::string deletionTimeStr =
+			    filename.substr(filename.find_last_of('.') + 1);
+			if (!isValidTimestampFormat(deletionTimeStr)) {
+				safs_pretty_syslog(
+				    LOG_ERR, "Invalid timestamp format in file: %s, skipping.",
+				    file.path().string().c_str());
+				continue;
+			}
+			int errorCode;
+			const std::time_t deletionTime =
+			    getTimeFromString(deletionTimeStr, errorCode);
+			if (errorCode != SAUNAFS_STATUS_OK) {
+				safs_pretty_syslog(LOG_ERR,
+				                   "Failed to parse deletion time "
+				                   "from file: %s, skipping.",
+				                   file.path().string().c_str());
+				continue;
+			}
+			trashIndex->add(deletionTime, file.path().string(), diskPath);
+		}
+	}
+
+	return SAUNAFS_STATUS_OK;
+}
+
+bool ChunkTrashManagerImpl::isValidTimestampFormat(
+    const std::string &timestamp) {
+	return timestamp.size() == kTimeStampLength &&
+	       std::all_of(timestamp.begin(), timestamp.end(), ::isdigit);
+}
+
+void ChunkTrashManagerImpl::removeExpiredFiles(const time_t &timeLimit,
+                                               size_t bulkSize) {
+	const auto expiredFilesCollection =
+	    trashIndex->getExpiredFiles(timeLimit, bulkSize);
+	removeTrashFiles(expiredFilesCollection);
+}
+
+void ChunkTrashManagerImpl::cleanEmptyFolders(const std::string &directory) {
+	if (!fs::exists(directory) || !fs::is_directory(directory)) {
+		return;  // Invalid path, so we do nothing
+	}
+
+	// Iterate through the directory's contents
+	for (const auto &entry : fs::directory_iterator(directory)) {
+		if (fs::is_directory(entry.path())) {
+			// Recursively clean subdirectories
+			cleanEmptyFolders(entry.path());
+		}
+	}
+
+	if (!isTrashPath(directory)) {
+		return;  // We only clean up trash directories
+	}
```

Why would it be anything other than isTrashPath? Seems like an assert would fit here

---

It is critical not to allow this component (chunk trash manager) to remove anything outside their folder. Regarding assert, it depends on someone not aware using `CMAKE_BUILD_TYPE=Release`, or removing the  
```
string(REPLACE "-DNDEBUG" "" CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO}")
```
line from the config (which is far away from this code), for the assert not to be available on release anymore. I'd rather not to take that risk. (assert is not meant for release, btw).

---

Could you not add both assert and return?

---

Or maybe throw an exception

---

Regardless, if a programmer calls this function, then a assert is better because when the programmer is testing, it will be likely he will find an assert. And calling this function is a programmer error, not a user error.

Also, I do know asserts work in production. I was recently testing migration and chunkserver asserted on a wrong chunk filename.

---

At the very least, it should not be a silent error, but sirens blaring because something is terribly wrong.

---

> ...we have asserts that work in production

Because of the `-DNDEBUG` line I mentioned above.

> it should not be a silent error

I agree, tx. It is no silent anymore, a log is generated now and an error code is returned.